cli/tools/registry/pm.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+use std::collections::HashMap;
+use std::path::PathBuf;
+
+use deno_ast::TextChange;
+use deno_config::FmtOptionsConfig;
+use deno_core::anyhow::bail;
+use deno_core::anyhow::Context;
+use deno_core::error::AnyError;
+use deno_core::futures::FutureExt;
+use deno_core::futures::StreamExt;
+use deno_core::serde_json;
+use deno_runtime::deno_fetch::reqwest;
+use deno_semver::jsr::JsrPackageReqReference;
+use deno_semver::npm::NpmPackageReqReference;
+use deno_semver::package::PackageReq;
+use jsonc_parser::ast::ObjectProp;
+use jsonc_parser::ast::Value;
+
+use super::api;
+use crate::args::jsr_api_url;
+use crate::args::AddFlags;
+use crate::args::Flags;
+use crate::factory::CliFactory;
+
+struct SelectedPackage {
+  import_name: String,
+  package_name: String,
+  version: String,
+}
+
+enum PackageAndVersion {
+  NotFound(String),
+  Selected(SelectedPackage),
+}
+
+async fn jsr_find_package_and_select_version(
+  client: &reqwest::Client,
+  registry_api_url: &str,
+  req: &PackageReq,
+) -> Result<PackageAndVersion, AnyError> {
+  let jsr_prefixed_name = format!("jsr:{}", req.name);
+
+  // TODO(bartlomieju): Need to do semver as well - @luca/flag@^0.14 should use to
+  // highest possible `0.14.x` version.
+  let version_req = req.version_req.version_text();
+  if version_req != "*" {
+    bail!("Specifying version constraints is currently not supported. Package: {}@{}", jsr_prefixed_name, version_req);
+  }
```

Comment:
It would maybe be nice if the code used the same infrastructure that the jsr specifier completions uses in the lsp (but maybe with cache busting that happens on each run, i forget how it works). Then this would be more easily implemented. Maybe that could be done after https://github.com/denoland/deno/pull/22612 lands

Reply:
I used code from @22612, but still left this part. We want to actually compute the best matching semver version here which I believe `PackageSearchApi` doesn't do.